token-lending/program/tests/modify_reserve_config.rs

```diff
@@ -0,0 +1,380 @@
+        .process_transaction(transaction)
+        .await
+        .map_err(|e| e.unwrap());
+
+    assert_eq!(
+        result.unwrap_err(),
+        TransactionError::InstructionError(
+            0,
+            InstructionError::Custom(LendingError::InvalidAccountInput as u32)
+        )
+    );
+
+    let reserve_info = sol_test_reserve.get_state(&mut banks_client).await;
+    assert_eq!(
+        reserve_info.config.optimal_utilization_rate,
+        TEST_RESERVE_CONFIG.optimal_utilization_rate
+    );
+}
+
+#[tokio::test]
+async fn correct_owner_providing_wrong_lending_market_fails() {
+    // When the correct owner of the lending market and reserve provides, perhaps inadvertently,
+    // a lending market that is different from the given reserve's corresponding lending market,
+    // then the transaction to modify the current reserve config should fail.
+    let mut test = ProgramTest::new(
+        "spl_token_lending",
+        spl_token_lending::id(),
+        processor!(process_instruction),
+    );
+
+    test.set_compute_max_units(70_000);
+
+    let user_accounts_owner = Keypair::new();
+    let lending_market = add_lending_market(&mut test);
+    let sol_oracle = add_sol_oracle(&mut test);
+
+    const SOL_DEPOSIT_AMOUNT_LAMPORTS: u64 = 10 * LAMPORTS_TO_SOL * INITIAL_COLLATERAL_RATIO;
+    const SOL_RESERVE_COLLATERAL_LAMPORTS: u64 = 2 * SOL_DEPOSIT_AMOUNT_LAMPORTS;
+
+    let sol_test_reserve = add_reserve(
+        &mut test,
+        &lending_market,
+        &sol_oracle,
+        &user_accounts_owner,
+        AddReserveArgs {
+            user_liquidity_amount: SOL_RESERVE_COLLATERAL_LAMPORTS,
+            liquidity_amount: SOL_RESERVE_COLLATERAL_LAMPORTS,
+            liquidity_mint_decimals: 9,
+            liquidity_mint_pubkey: spl_token::native_mint::id(),
+            config: TEST_RESERVE_CONFIG,
+            mark_fresh: true,
+            ..AddReserveArgs::default()
+        },
+    );
+
+    let other_lending_market = add_lending_market(&mut test);
+
+    let (mut banks_client, payer, recent_blockhash) = test.start().await;
+
+    const OPTIMAL_UTILIZATION_RATE_CHANGE: u8 = 10;
+
+    let new_config = ReserveConfig {
+        optimal_utilization_rate: TEST_RESERVE_CONFIG.optimal_utilization_rate
+            - OPTIMAL_UTILIZATION_RATE_CHANGE,
+        loan_to_value_ratio: 50,
+        liquidation_bonus: 5,
+        liquidation_threshold: 55,
+        min_borrow_rate: 0,
+        optimal_borrow_rate: 4,
+        max_borrow_rate: 30,
+        fees: ReserveFees {
+            borrow_fee_wad: 100_000_000_000,
+            flash_loan_fee_wad: 3_000_000_000_000_000,
+            host_fee_percentage: 20,
+        },
+    };
+
+    let mut transaction = Transaction::new_with_payer(
+        &[modify_reserve_config(
+            spl_token_lending::id(),
+            new_config,
+            sol_test_reserve.pubkey,
+            other_lending_market.pubkey,
+            other_lending_market.owner.pubkey(),
```

Comment:
If you want to test wrong lending market with correct owner, shouldn't this be `lending_market.owner.pubkey()`?  Then you can trip `InvalidMarketOwner` in your test.

Otherwise, is this exactly the same as the previous test?

Reply:
So `lending_market.owner.pubkey() == other_lending_market.owner.pubkey()` due to the fact that the `add_lending_market` helper, which creates both lending markets in this test, uses a [static keypair file](https://github.com/solana-labs/solana-program-library/blob/d1e842ca3f5e1752a2e436e711d9fa09b486f119/token-lending/program/tests/helpers/mod.rs#L96) as the owner for every invocation. The error would still be `InvalidAccountInput` since the owner is indeed valid (the same) but the lending market is not.

But I agree with you - if line 355 in the above just uses `lending_market.owner.pubkey()` as the argument, then the test case of "correct owner but incorrect market" would be more clear. I'll make this change. Thank you!

Note: The previous test, `owner_of_different_lending_market...` should be testing the `InvalidMarketOwner` error. I'll modify the previous test accordingly.

Reply:
To summarize, the four (updated) tests should test:
1. happy path - right market, right owner can change ReserveConfig
2. invalid signer - right market, right owner but owner is not a signer -> error `InvalidSigner`
3. invalid owner - right market, wrong owner -> error `InvalidMarketOwner`
4. invalid market - wrong market, right owner -> error `InvalidAccountInput`

Reply:
This is all perfect, well done!